### ISSUE TITLE
filter=lfs diff=lfs merge=lfs

### ISSUE THREAD

**Issue by rygo6** [github/git-lfs#604]
I am finding the documentation for LFS to be somewhat lacking, so I am opening issues on what I can't find in the documentation.

Right now I am wondering about, and a bit confused about the resulting addition git lfs track adds to the gitattibutes file.

For the command:

    git lfs track "*.psd"

It is adding into .gitattributes

    *.psd filter=lfs diff=lfs merge=lfs -text

So I presume filter is referring to the [filter "lfs"] in the .gitconfig.

But there is a lfs diff and a lfs merge? Where are these defined? What are these?

Does this mean I can have an lfs filter, an lfs diff but then merge those files with something else?

For example, I am using git on Unity3D projects. Unity has a YAMLMerge tool for it's scenes which are stored in YAML. http://docs.unity3d.com/Manual/SmartMerge.html

Is it possible for me to filter and diff the YAML scene files with lfs, but then specify the YAMLMerge tool as the merge attribute? Will this set up know to merge the actual YAML files, or will it try to end up merging the LFS pointer files?

**Comment by technoweenie** [comment]
Git LFS doesn't have a custom merge or diff tool right now. Those hooks are reserved for future use. However, you can edit the `.gitattributes` to do whatever you like. As long as the `filter=lfs` setting is kept, Git LFS will work. It may even be a good idea to allow this in the `track` command:

    $ git lfs track "*.psd" --filter=unityyamlmerge

How big are these scene files? Yaml is plain text, so it makes sense to use regular git for them. Usually you'd reserve larger binary files for Git LFS.

**Comment by strich** [comment]
FYI the Unity scene files can be 10-100+ MB in size, if the user sets Unity to serialize scene files to plain text.

**Comment by rygo6** [comment]
@technoweenie I think I may be misunderstanding something, because I would expect the command to be:

    $ git lfs track "*.unity" --merge=unityyamlmerge

To then add in

   *.unity filter=lfs diff=lfs merge=unityyamlmerge -text

?

Also, as strich said. Unity YAML scene files can 100mb or more. They are also files that the user does not really need to see, on GitHub or other. So having them be treated as binary and put onto LFS may be a good idea.

What I was curious to know is. If I put a merge value on an entry which has filter=lfs. Does it know to merge the actual files, and not the text pointer files?

**Comment by technoweenie** [comment]
@rygo6 Oh good call, a `--merge` flag would be better. You can also just update the entries in `.gitattributes` yourself. Git LFS won't rewrite it or anything, unless you untrack and track again.

> What I was curious to know is. If I put a merge value on an entry which has filter=lfs. Does it know to merge the actual files, and not the text pointer files?

The text pointer files.

**Comment by rygo6** [comment]
What then is the command to merge the real files and not the pointer files?

**Comment by technoweenie** [comment]
There isn't one, currently. Git LFS is designed for binary objects that aren't mergeable from the command line. However, the `merge=lfs` property in the `.gitattributes` file gives us a hook to build this out in the future. Preferably by someone that knows Unity already. We're still working on core Git LFS features, so it may be a little while before we get to tool specific things like this.

**Comment by rygo6** [comment]
Just so I can understand better. If I wanted to write something to merge files stored on LFS. I would need to go through a process like this:

1. Retrieve the real files associated with the two pointer files in conflict.
2. Merge those two files.
3. Remove the conflicted pointer file from head.
4. Save the merged file to the head.

Then add and commit that merged filed to resolve the conflict. LFS automatically making the pointer file and dealing with it that way.

Does that all sound correct?

Also, I am interested in this not just to merge .unity files through yamlmerge. But I am also interested in writing a tool to help merge png files and also FBX files through some kind of visual comparing system.

The point which is a grey area for me is point 1. Is there a command to grab a file associated with a pointer file?

**Comment by technoweenie** [comment]
Actually, I'm actually not well versed in how merge tools are implemented. If they happen to work on files in the working directory, it's possible that they should "just work" without any LFS specific commands. I'm assuming that they work based on the git blob contents, which contain the LFS pointer.

Git LFS uses the clean and smudge filters to convert files back and forth. They follow Git's internal `.gitattributes` API, which uses STDIN and STDOUT.

If you have a pointer, but you want the blob, you'll want the smudge command. Write the pointer contents to STDIN, and the command will output the actual file to STDOUT. You can simulate this with `echo`:

```bash
$ echo "version https://git-lfs.github.com/spec/v1
oid sha256:55d51edb307ed7bf4a1bbca3867d132a07f76828d57f37e0f31f712c02ceafe0
size 48202" | git lfs smudge > some-file
```

You'll probably want to redirect this to some temporary file. This command will not download it if it already exists in the `.git/lfs/objects` directory.

The clean filter is basically the reverse. Write the file contents to STDIN, and it spits out the pointer file to STDOUT.

```bash
# assume that the above command wrote to some-file
$ cat some-file | git lfs clean
version https://git-lfs.github.com/spec/v1
oid sha256:55d51edb307ed7bf4a1bbca3867d132a07f76828d57f37e0f31f712c02ceafe0
size 48202
```

Your merge tool will want to write that file as the git blob for that file's git tree entry. The clean command will automatically put that file into `.git/lfs/objects` to be uploaded on the next `git push`.

I do plan to dive into how merge tools work eventually. We can add whatever convenience commands to make this easier.

**Comment by sinbad** [comment]
When merging git writes 3 temp files of the 3 source states: base (common ancestor), local (current branch) and remote (branch being merged), then calls the merge tool passing the paths to these 3 files, plus a path where it should write the merged result.

I think we should make a general git-lfs merge tool which is just a conduit for other merge tools that would normally plug directly into .gitattributes, but which we need to redirect to the real content. So when the lfks merge tool gets called, it reads the contents of base/local/remote to get the SHAs, then copies the content from .git/lfs/objects to 3 *other* temp files, and creates a 4th temp file for the merge content, then calls the standard merge tool (like Unity yamlmerge) with these paths. 

I'm not sure if the clean filter is called on the merge output, if so that's basically it (we can let the actual merge tool write to the original merge output path as given by git), if not then we'd need to convert the merge output to a pointer, write the content to the lfks store then give the pointer back to git merge.

**Comment by callumtw** [comment]
I have exactly the same goal: put Unity scene files on LFS but merge them with Unity's Smart Merge tool. 

Most people using Git LFS with Unity seem to be leaving Unity scene files to be handled by regular git, specifying `merge=unityyamlmerge`, but I think this is simply out of uncertainty whether the two can be used together rather than out of preference. Despite being text, unity scene files can get much, much larger than other LFS'd assets, so it makes sense to treat them similarly. I was hoping now several years have passed the situation may have developed?


In short, is it now possible to track a (text) file type on LFS but specify a custom merge tool? E.g the following, or an equivalent:

```
*.unity filter=lfs diff=lfs merge=unityyamlmerge -text
```

From the above discussion, it sounds like this might attempt to use the tool to merge the LFS pointers, not the actual files.

Apologies if this has since been documented, but this was the only relevant page I could find. However it looks like it may be related to #4820 ? Sadly I don't know enough git lingo to be sure if a merge driver is exactly what's needed here.

**Comment by xuchys-jeffyP-gewvi0** [comment]
Y
